File: cpp/Crud.cpp
// import library dan file
#include <bits/stdc++.h>
#include "Mahasiswa.cpp"

// using standard namespace
using namespace std;

class Crud
{
    /* class crud berisi vector of object dari class mahasiswa
    yang digunakan untuk memodifikasi vector tersebut, seperti
    menambahkan data, menampilkan data, memperbaharui data,
    dan menghapus data*/

private:
    // private atribut
    vector<Mahasiswa> dataMhs;

public:
    // CONSTRUCTOR
    Crud(vector<Mahasiswa> data)
    {
        /* konstruktor yang menerima parameter dengan tipe data vector sebagai inputannya*/
        dataMhs = data;
    }

    // PUBLIC METHOD
    int cariNIM(string nim)
    {
        /*method yang digunakan untuk mencari nim
        dan mengembalikan index nim tersebut*/
        for (int i = 0; i < dataMhs.size(); i++)
        {
            if (nim == dataMhs[i].getNIM()) // jika nim ditemukan
            {
                return i;
            }
        }
        return -1; // jika nim tidak ditemukan
    }

    void tampilData()
    {
        /*method yang digunakan untuk menampilkan data mahasiswa
        yang ada di dalam vector*/
        cout << "Data Mahasiswa : \n";
        for (int i = 0; i < dataMhs.size(); i++)
        {
            cout << "==========================================\n";
            cout << "Nama          : " << dataMhs[i].getNama() << endl;
            cout << "NIM           : " << dataMhs[i].getNIM() << endl;
            cout << "Program Studi : " << dataMhs[i].getProdi() << endl;
            cout << "Fakultas      : " << dataMhs[i].getFakultas() << endl;
        }
    }

    void tambahData(string nama, string nim, string prodi, string fakultas)
    {
        /*method yang digunakan untuk menambahkan data baru
        ke dalam vector*/
        Mahasiswa temp(nama, nim, prodi, fakultas); // instansiasi objek mahasiswa
        dataMhs.push_back(temp);                    // menambahkannya ke vector
    }

    void ubahData(int indexNIM, string nama, string nim, string prodi, string fakultas)
    {
        /*method yang digunakan untuk memperbaharui data mahasiswa
        pada index tertentu*/
        // memanggil setter untuk setiap atribut
        dataMhs[indexNIM].setNama(nama);
        dataMhs[indexNIM].setNIM(nim);
        dataMhs[indexNIM].setProdi(prodi);
        dataMhs[indexNIM].setFakultas(fakultas);
    }

    void hapusData(int indexNIM)
    {
        /*method yang digunakan untuk menghapus data mahasiswa
        pada index tertentu*/
        dataMhs.erase(dataMhs.begin() + indexNIM); // menggunakan erase() untuk menghapus data
    }

    // DESTRUCTOR
    ~Crud() {}
};

File: java/Crud.java

// import library
import java.util.ArrayList;

public class Crud {
    /*
     * class crud berisi arraylist of object dari class mahasiswa
     * yang digunakan untuk memodifikasi arraylist tersebut, seperti
     * menambahkan data, menampilkan data, memperbaharui data,
     * dan menghapus data
     */

    // private atribut
    private ArrayList<Mahasiswa> dataMhs;

    // CONSTRUCTOR
    Crud(ArrayList<Mahasiswa> data) {
        /*
         * konstruktor yang menerima parameter dengan tipe data arraylist sebagai
         * inputannya
         */
        dataMhs = data;
    }

    // PUBLIC METHOD
    int cariNIM(String nim) {
        /*
         * method yang digunakan untuk mencari nim
         * dan mengembalikan index nim tersebut
         */
        for (int i = 0; i < dataMhs.size(); i++) {
            if (dataMhs.get(i).getNIM().equals(nim)) {// jika nim ditemukan
                return i;
            }
        }
        return -1;// jika nim tidak ditemukan
    }

    void tambahData(String nama, String nim, String prodi, String fakultas) {
        /*
         * method yang digunakan untuk menambahkan data baru
         * ke dalam vector
         */

        Mahasiswa temp = new Mahasiswa(nama, nim, prodi, fakultas); // instansiasi objek mahasiswa
        dataMhs.add(temp); // menambahkan objek ke list
    }

    void tampilData() {
        System.out.println("Data Mahasiswa");
        for (int i = 0; i < dataMhs.size(); i++) {
            /*
             * method yang digunakan untuk menampilkan data mahasiswa
             * yang ada di dalam vector
             */
            System.out.println("==========================================");
            System.out.println("Nama          : " + dataMhs.get(i).getNama());
            System.out.println("NIM           : " + dataMhs.get(i).getNIM());
            System.out.println("Program Studi : " + dataMhs.get(i).getProdi());
            System.out.println("Fakultas      : " + dataMhs.get(i).getFakultas());
        }
    }

    void hapusData(int indexNIM) {
        /*
         * method yang digunakan untuk menghapus data mahasiswa
         * pada index tertentu
         */
        dataMhs.remove(indexNIM); // menggunakan remove() untuk menghapus data
    }

    void ubahData(int indexNIM, String nama, String nim, String prodi, String fakultas) {
        /*
         * method yang digunakan untuk memperbaharui data mahasiswa
         * pada index tertentu
         */
        // memanggil setter untuk setiap atribut
        dataMhs.get(indexNIM).setNama(nama);
        dataMhs.get(indexNIM).setNIM(nim);
        dataMhs.get(indexNIM).setProdi(prodi);
        dataMhs.get(indexNIM).setFakultas(fakultas);
    }

}
